### PR TITLE
KAFKA-14412: Transactional StateStores

### PR DESCRIPTION
**Pull request by nicktelford**
See KIP-892: Transactional Semantics for StateStores[1] for a detailed explanation of the design.

---

Writes to StateStores are now isolated from the underlying RocksDB database until the changelog transaction has been committed. We do this by writing first to a RocksDB `WriteBatchWithIndex`, which on-commit will be written to the RocksDB database.

Now that writes to a StateStore are isolated by a "transaction buffer" (`WriteBatchWithIndex`), we can also isolate them from interactive query threads. We introduce a new configuration,
`default.state.isolation.level`, which defaults to `READ_UNCOMMITTED` to provide the existing behaviour.

When `processing.mode` is set to `exactly-once(-v2|-beta)`, the default swtiches to `READ_COMMITTED`, as this will be the desired behaviour for the vast majority of use-cases.

Crucially, this ensures that in the event of an error, state stores no longer need to be wiped under EOS, unless the user has explicitly selected the `READ_UNCOMMITTED` isolation level.

---

To ensure consistency, we now store offsets in a secondary column-family in RocksDB (`__offsets`). These offsets are updated on-commit. We guarantee that the `__offsets` memtables and data memtables are flushed atomically, together, using RocksDB's Atomic Flush. This allows us to remove the explicit `flush()` calls that previously forcibly flushed memtables to disk. RocksDB now entirely controls memtable flush behaviour via its configuration.

We continue writing offsets to the `.checkpoint` file, to ensure that during rebalance, we can send offsets for currently closed stores, without having to open them just to read their offset. The `.checkpoint` file is only updated on store close for this reason. If the store is open, the currently in-memory offset is used instead.

We also keep the `.checkpoint` file for custom `StateStore` implementations that do not manage their own offsets. For these stores, the legacy behaviour will still be used.

---

The performance should be at least as good, if not better, than before, as batched writes to RocksDB avoid locking the memtable for every write.

1: https://cwiki.apache.org/confluence/display/KAFKA/KIP-892%3A+Transactional+Semantics+for+StateStores

### CODE REVIEW

**Comment by nicktelford** [comment]
@cadonna @mjsax @ableegoldman @lucasbru @wcarlson5 @bbejeck @vvcephei @guozhangwang 

My apologies to whoever draws the short straw of having to review this!

This is the main implementation of KIP-892. I broke out as much as I could into independent PRs that have been merged ahead of this.

If you would like to discuss anything in-person, I'm available on the Confluent Community Slack between 10:00 and 17:00 UTC, Monday to Friday. Happy to discuss on a video call if necessary.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java [hunk 1]
```diff
@@ -262,8 +268,7 @@ private void closeDirtyAndRevive(final Collection<Task> taskWithChangelogs, fina
                 final Collection<TopicPartition> corruptedPartitions = task.changelogPartitions();
 
                 // mark corrupted partitions to not be checkpointed, and then close the task as dirty
-                // TODO: this step should be removed as we complete migrating to state updater
-                if (markAsCorrupted && stateUpdater == null) {
+                if (markAsCorrupted) {
```
Comment (nicktelford) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:271]: This is necessary now that we only wipe stores when `store.corrupted == true`.

File: checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -135,7 +135,7 @@
     </module>
     <module name="CyclomaticComplexity">
       <!-- default is 10-->
-      <property name="max" value="16"/>
+      <property name="max" value="17"/>
```
Comment (nicktelford) [inline checkstyle/checkstyle.xml:138]: `TaskExecutor#commitOffsetsOrTransaction`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 3]
```diff
@@ -266,29 +269,20 @@ void initializeStoreOffsetsFromCheckpoint(final boolean storeDirIsEmpty) {
...
+                    // load offsets from .checkpoint file
+                    offset = loadOffsetFromCheckpointFile(offset, loadedCheckpoints, store);
+
+                    // no offsets found for store, store is corrupt if not empty
+                    throwCorruptIfNoOffsetAndNotEmpty(offset, storeDirIsEmpty);
+
+                    updateOffsetInMemory(offset, store);
+
+                    syncManagedOffsetInStore(offset, store);
+
```
Comment (nicktelford) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:285]: I generally tried to avoid unnecessary changes/refactoring wherever possible, but the new logic here caused checkstyle to complain about NPath complexity, so I factored it out to several `private` methods.

In the end, breaking this up has made it considerably easier to follow, and is now somewhat self-documenting.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 4]
```diff
@@ -805,6 +959,194 @@ public void reset() {
...
+                          final ColumnFamilyHandle offsetsCF,
+                          final WriteOptions writeOptions) {
+            this.db = db;
+            this.offsetsCF = offsetsCF;
+            this.writeOptions = writeOptions;
+        }
+
+        @Override
+        public byte[] get(final ColumnFamilyHandle columnFamily, final byte[] key) throws RocksDBException {
+            if (Thread.currentThread() instanceof ProcessingThread || isStreamThreadForTest) {
```
Comment (nicktelford) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:1025]: We need to do this to ensure that, under `READ_COMMITTED`, interactive query threads don't read from the transaction buffer, since it's not thread-safe.

While it's not ideal to add more to the hot path, `Thread.currentThread()` and `instanceof` checks are both JVM intrinsics, so should be negligible compared with the cost of querying RocksDB.

**Comment by nicktelford** [comment]
The build says there are "19 new failed tests", but they all pass locally. Since all but 3 are in non-streams modules, this looks like poor CI performance induced timeouts.

**Comment by ableegoldman** [comment]
Hey @nicktelford , I'll try to take a look at this here and there and I can't speak for the Confluent folks, but I have to say it's pretty intimidating to see a PR with 3000 LOC and over 100 files changed...especially one that touches on some complicated stuff that's fundamental to Kafka Streams. 

I know you've already tried to break this work up into multiple PRs but for your own sake as much as ours, can you take another look at whether you can peel off any more into separate PRs? I'm just worried that it's going to be really hard for this to make progress as such a large PR, and the longer it's stalled the harder it'll be to keep it updated.

**Comment by nicktelford** [comment]
> Hey @nicktelford , I'll try to take a look at this here and there and I can't speak for the Confluent folks, but I have to say it's pretty intimidating to see a PR with 3000 LOC and over 100 files changed...especially one that touches on some complicated stuff that's fundamental to Kafka Streams.
> 
> I know you've already tried to break this work up into multiple PRs but for your own sake as much as ours, can you take another look at whether you can peel off any more into separate PRs? I'm just worried that it's going to be really hard for this to make progress as such a large PR, and the longer it's stalled the harder it'll be to keep it updated.

Hi @ableegoldman, I totally understand, which is why I already peeled some out. My general policy with PRs is that they must be committable independently, without breaking the build or cause performance regressions. I'll do another pass over this to see if I can break a bit more out, but I can't imagine it'll be more than a few LoC here and there.

If you would accept PRs that cannot be merged independently without breaking the build, or perhaps use a staging branch to integrate the changes piece-by-piece, then I can definitely break it up some more.

FWIW, approximately half of all the changes are in `streams/src/test`.

**Comment by cadonna** [comment]
Some suggestions/questions to split the PR in smaller parts (without having looked at the PR in much depth):
- Is it possible to have a separate PR for the config changes?
- Is it possible to add code without removing existing code? Maybe behind an internal flag that can later be removed in a PR.

**Comment by nicktelford** [comment]
> * Is it possible to have a separate PR for the config changes?

@cadonna Yes, in fact I'm working on that as I speak. :grin: It'll be a little strange to essentially add config that "does nothing", but I think it's acceptable, provided it doesn't alter behaviour (yet).

> Is it possible to add code without removing existing code? Maybe behind an internal flag that can later be removed in a PR.

I think the main issue is that a large amount of the changes _depend_ on others, so in many cases it won't make sense to include one without the other, even behind a flag. I'll see what I can do though.

**Comment by nicktelford** [comment]
Closing this PR for now while I break changes out into separate PRs
